False error in conditional expression returning union

* Are you reporting a bug, or opening a feature request?

Bug

* Please insert below the code you are checking with mypy,
  or a mock-up repro if the source is private. We would appreciate
  if you try to simplify your case to a minimal repro.

```py
from typing import TypeVar, Union


T = TypeVar("T")


def get_or_str(value: T) -> Union[str, T]:
    return value


def f(b: bool) -> Union[str, bool]:
    return "" if b else get_or_str(True)
```

* What is the actual behavior/output?

`a.py:12: error: Argument 1 to "get_or_str" has incompatible type "bool"; expected "str"`

* What is the behavior/output you expect?

No error.

* What are the versions of mypy and Python you are using?
  Do you see the same issue after installing mypy from Git master?

0.710+dev.5f08ccf029aa3046b15e1afc60743ba692e4758d (from master)

* What are the mypy flags you are using? (For example --strict-optional)

None

There is a similar error when trying to return a `Sequence` using tuples of different lengths:

```py
def f(b: bool) -> Sequence[int]:
    return (1,) if b else (2, 3)
```

`a.py:4: error: Incompatible return value type (got "object", expected "Sequence[int]")`

A similar problem:
```py
from typing import Union

def f(a: Union[str, int], b: Union[str, int]) -> Union[str, float]:
    return a if isinstance(a, str) else b if isinstance(b, str) else a / b
```

`a.py:4: error: Incompatible return value type (got "object", expected "Union[str, int]")`

Likely the same issue:
```py
from typing import Callable, Optional, TypeVar

T = TypeVar("T")

def f(
    a: Callable[[], T],
    b: Callable[[], T]
) -> T:
    return a()

op: Optional[int] = f(lambda: 1, lambda: None)
```

`a.py:11: error: Incompatible return value type (got "None", expected "int")`

Works bare `T` instead of callables.

--- a/mypy/checkexpr.py
+++ b/mypy/checkexpr.py
@@ -5879,10 +5879,16 @@ class ExpressionChecker(ExpressionVisitor[Type], ExpressionCheckerSharedApi):
         allow_none_return: bool = False,
         suppress_unreachable_errors: bool = True,
     ) -> Type:
+        ctx_items = []
+        if context is not None:
+            ctx_items.append(context)
         if self.type_context and self.type_context[-1] is not None:
-            ctx = make_simplified_union([context, self.type_context[-1]])
+            ctx_items.append(self.type_context[-1])
+        ctx: Type | None
+        if ctx_items:
+            ctx = make_simplified_union(ctx_items)
         else:
-            ctx = context
+            ctx = None
         with self.chk.binder.frame_context(can_skip=True, fall_through=0):
             if map is None:
                 # We still need to type check node, in case we want to

--- a/mypy/checkexpr.py
+++ b/mypy/checkexpr.py
@@ -5879,16 +5879,20 @@ class ExpressionChecker(ExpressionVisitor[Type], ExpressionCheckerSharedApi):
         allow_none_return: bool = False,
         suppress_unreachable_errors: bool = True,
     ) -> Type:
+        if self.type_context and self.type_context[-1] is not None:
+            ctx = make_simplified_union([context, self.type_context[-1]])
+        else:
+            ctx = context
         with self.chk.binder.frame_context(can_skip=True, fall_through=0):
             if map is None:
                 # We still need to type check node, in case we want to
                 # process it for isinstance checks later. Since the branch was
                 # determined to be unreachable, any errors should be suppressed.
                 with self.msg.filter_errors(filter_errors=suppress_unreachable_errors):
-                    self.accept(node, type_context=context, allow_none_return=allow_none_return)
+                    self.accept(node, type_context=ctx, allow_none_return=allow_none_return)
                 return UninhabitedType()
             self.chk.push_type_map(map)
-            return self.accept(node, type_context=context, allow_none_return=allow_none_return)
+            return self.accept(node, type_context=ctx, allow_none_return=allow_none_return)
 
     #
     # Helpers

--- a/mypy/checkexpr.py
+++ b/mypy/checkexpr.py
@@ -4271,7 +4271,9 @@ class ExpressionChecker(ExpressionVisitor[Type], ExpressionCheckerSharedApi):
         ):
             self.msg.unreachable_right_operand(e.op, e.right)
 
-        right_type = self.analyze_cond_branch(right_map, e.right, expanded_left_type)
+        right_type = self.analyze_cond_branch(
+            right_map, e.right, self._combined_context(expanded_left_type)
+        )
 
         if left_map is None and right_map is None:
             return UninhabitedType()
@@ -5879,26 +5881,26 @@ class ExpressionChecker(ExpressionVisitor[Type], ExpressionCheckerSharedApi):
         allow_none_return: bool = False,
         suppress_unreachable_errors: bool = True,
     ) -> Type:
-        ctx_items = []
-        if context is not None:
-            ctx_items.append(context)
-        if self.type_context and self.type_context[-1] is not None:
-            ctx_items.append(self.type_context[-1])
-        ctx: Type | None
-        if ctx_items:
-            ctx = make_simplified_union(ctx_items)
-        else:
-            ctx = None
         with self.chk.binder.frame_context(can_skip=True, fall_through=0):
             if map is None:
                 # We still need to type check node, in case we want to
                 # process it for isinstance checks later. Since the branch was
                 # determined to be unreachable, any errors should be suppressed.
                 with self.msg.filter_errors(filter_errors=suppress_unreachable_errors):
-                    self.accept(node, type_context=ctx, allow_none_return=allow_none_return)
+                    self.accept(node, type_context=context, allow_none_return=allow_none_return)
                 return UninhabitedType()
             self.chk.push_type_map(map)
-            return self.accept(node, type_context=ctx, allow_none_return=allow_none_return)
+            return self.accept(node, type_context=context, allow_none_return=allow_none_return)
+
+    def _combined_context(self, ty: Type | None) -> Type | None:
+        ctx_items = []
+        if ty is not None:
+            ctx_items.append(ty)
+        if self.type_context and self.type_context[-1] is not None:
+            ctx_items.append(self.type_context[-1])
+        if ctx_items:
+            return make_simplified_union(ctx_items)
+        return None
 
     #
     # Helpers

--- a/test-data/unit/check-inference-context.test
+++ b/test-data/unit/check-inference-context.test
@@ -1510,3 +1510,24 @@ def mymin(
 def check(paths: Iterable[str], key: Callable[[str], int]) -> Union[str, None]:
     return mymin(paths, key=key, default=None)
 [builtins fixtures/tuple.pyi]
+
+[case testBinaryOpInferenceContext]
+from typing import Literal, TypeVar
+
+T = TypeVar("T")
+
+def identity(x: T) -> T:
+    return x
+
+def check1(use: bool, val: str) -> "str | Literal[True]":
+    return use or identity(val)
+
+def check2(use: bool, val: str) -> "str | bool":
+    return use or identity(val)
+
+def check3(use: bool, val: str) -> "str | Literal[False]":
+    return use and identity(val)
+
+def check4(use: bool, val: str) -> "str | bool":
+    return use and identity(val)
+[builtins fixtures/tuple.pyi]

